File: backend/routes/payments.py
"""Payment routes using PhonePe v2 OAuth + webhook (signature-validated)."""
import base64
import json
import logging
import uuid
from datetime import datetime, timezone
from typing import Any, Dict

import httpx
from bson import ObjectId
from fastapi import APIRouter, HTTPException, Depends, Request

from core.config import FRONTEND_URL, PHONEPE_ENV, PHONEPE_PAY_URL, PHONEPE_STATUS_BASE
from core.database import db
from core.deps import get_current_user
from core.models import PaymentInitiateRequest
from services.activity import log_activity
from services.phonepe import (
    get_phonepe_token,
    verify_webhook_auth,
    webhook_auth_required,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/payments", tags=["payments"])


@router.post("/initiate")
async def initiate_payment(body: PaymentInitiateRequest, current_user: dict = Depends(get_current_user)):
    try:
        appt = await db.appointments.find_one({"_id": ObjectId(body.appointment_id)})
    except Exception:
        raise HTTPException(status_code=404, detail="Appointment not found")
    if not appt:
        raise HTTPException(status_code=404, detail="Appointment not found")
    if appt["user_id"] != current_user["_id"]:
        raise HTTPException(status_code=403, detail="Access denied")

    amount_paise = int(appt.get("consultation_fee", 2000) * 100)
    merchant_order_id = f"MC{str(uuid.uuid4())[:8].upper()}"
    txn_id = merchant_order_id
    redirect_url = f"{FRONTEND_URL}/payment/status?txnId={txn_id}"

    checkout_url = None
    api_error = None
    access_token = await get_phonepe_token()
    if access_token:
        try:
            payload = {
                "merchantOrderId": merchant_order_id,
                "amount": amount_paise,
                "expireAfter": 1200,
                "metaInfo": {
                    "udf1": current_user["_id"],
                    "udf2": appt.get("doctor_name", ""),
                    "udf3": str(appt["_id"]),
                },
                "paymentFlow": {
                    "type": "PG_CHECKOUT",
                    "message": f"Consultation fee - {appt.get('doctor_name', '')}",
                    "merchantUrls": {"redirectUrl": redirect_url},
                },
            }
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"O-Bearer {access_token}",
            }
            async with httpx.AsyncClient(timeout=10.0) as c:
                resp = await c.post(PHONEPE_PAY_URL, json=payload, headers=headers)
            if resp.status_code in (200, 201):
                data = resp.json()
                checkout_url = data.get("redirectUrl") or data.get("data", {}).get("redirectUrl")
            else:
                api_error = f"{resp.status_code}: {resp.text[:200]}"
                logger.warning(f"PhonePe pay init failed: {api_error}")
        except Exception as e:
            api_error = str(e)
            logger.warning(f"PhonePe API exception: {e}")

    if not checkout_url:
        checkout_url = f"{FRONTEND_URL}/payment/simulate/{txn_id}"

    await db.transactions.insert_one({
        "appointment_id": str(appt["_id"]),
        "user_id": current_user["_id"],
        "merchant_order_id": merchant_order_id,
        "transaction_id": txn_id,
        "amount": amount_paise,
        "payment_state": "PENDING",
        "checkout_url": checkout_url,
        "doctor_name": appt.get("doctor_name", ""),
        "appointment_date": appt.get("appointment_date", ""),
        "appointment_time": appt.get("appointment_time", ""),
        "phonepe_env": PHONEPE_ENV,
        "api_error": api_error,
        "created_at": datetime.now(timezone.utc),
        "updated_at": datetime.now(timezone.utc),
    })
    await db.appointments.update_one(
        {"_id": ObjectId(body.appointment_id)},
        {"$set": {"transaction_id": txn_id}},
    )
    return {
        "checkout_url": checkout_url,
        "transaction_id": txn_id,
        "merchant_order_id": merchant_order_id,
        "is_simulation": "/simulate/" in checkout_url,
    }


@router.get("/status/{txn_id}")
async def get_payment_status(txn_id: str):
    txn = await db.transactions.find_one({"transaction_id": txn_id})
    if not txn:
        raise HTTPException(status_code=404, detail="Transaction not found")
    txn["_id"] = str(txn["_id"])
    if txn["payment_state"] == "PENDING":
        access_token = await get_phonepe_token()
        if access_token:
            try:
                url = f"{PHONEPE_STATUS_BASE}/{txn['merchant_order_id']}/status?details=false&errorContext=true"
                async with httpx.AsyncClient(timeout=10.0) as c:
                    resp = await c.get(url, headers={
                        "Content-Type": "application/json",
                        "Authorization": f"O-Bearer {access_token}",
                    })
                if resp.status_code == 200:
                    data = resp.json()
                    new_state = data.get("state") or data.get("data", {}).get("state")
                    if new_state in ["COMPLETED", "FAILED"]:
                        await db.transactions.update_one(
                            {"transaction_id": txn_id},
                            {"$set": {"payment_state": new_state, "updated_at": datetime.now(timezone.utc)}},
                        )
                        txn["payment_state"] = new_state
                        if new_state == "COMPLETED":
                            await db.appointments.update_one(
                                {"transaction_id": txn_id},
                                {"$set": {"status": "confirmed", "payment_status": "paid"}},
                            )
                        else:
                            await db.appointments.update_one(
                                {"transaction_id": txn_id},
                                {"$set": {"status": "cancelled", "payment_status": "failed"}},
                            )
            except Exception as e:
                logger.warning(f"PhonePe status check failed: {e}")
    return txn


@router.post("/simulate/{txn_id}/success")
async def simulate_success(txn_id: str):
    txn = await db.transactions.find_one({"transaction_id": txn_id})
    if not txn:
        raise HTTPException(status_code=404, detail="Transaction not found")
    prov_ref = f"SIM{str(uuid.uuid4())[:8].upper()}"
    await db.transactions.update_one(
        {"transaction_id": txn_id},
        {"$set": {
            "payment_state": "COMPLETED", "payment_mode": "UPI_SIMULATE",
            "provider_reference_id": prov_ref, "updated_at": datetime.now(timezone.utc),
        }},
    )
    appt = await db.appointments.find_one({"transaction_id": txn_id})
    if appt:
        await db.appointments.update_one(
            {"transaction_id": txn_id},
            {"$set": {"status": "confirmed", "payment_status": "paid"}},
        )
        await log_activity(
            txn["user_id"], "Patient", "PAYMENT_SUCCESS",
            f"Appointment confirmed for {appt.get('doctor_name', '')}",
        )
    return {"message": "Payment simulated successfully", "transaction_id": txn_id, "provider_reference_id": prov_ref}


@router.post("/simulate/{txn_id}/failure")
async def simulate_failure(txn_id: str):
    txn = await db.transactions.find_one({"transaction_id": txn_id})
    if not txn:
        raise HTTPException(status_code=404, detail="Transaction not found")
    await db.transactions.update_one(
        {"transaction_id": txn_id},
        {"$set": {"payment_state": "FAILED", "updated_at": datetime.now(timezone.utc)}},
    )
    await db.appointments.update_one(
        {"transaction_id": txn_id},
        {"$set": {"status": "cancelled", "payment_status": "failed"}},
    )
    return {"message": "Payment failed", "transaction_id": txn_id}


@router.post("/webhook")
async def phonepe_webhook(request: Request):
    """
    PhonePe v2 webhook with signature validation.

    When PHONEPE_WEBHOOK_USERNAME and PHONEPE_WEBHOOK_PASSWORD are set in env,
    incoming webhooks must include `Authorization: <SHA256(username:password)>`.
    When either is empty (dev mode), validation is skipped (logged as warning).

    Always responds 200 to acknowledge receipt (prevents infinite retries).
    """
    auth_header = request.headers.get("Authorization", "")
    if webhook_auth_required() and not verify_webhook_auth(auth_header):
        logger.warning("PhonePe webhook rejected: invalid Authorization signature")
        raise HTTPException(status_code=401, detail="Invalid webhook signature")
    if not webhook_auth_required():
        logger.warning("PhonePe webhook auth is DISABLED (dev mode). Set PHONEPE_WEBHOOK_USERNAME/PASSWORD before production.")

    try:
        raw = await request.body()
        body: Dict[str, Any] = {}
        try:
            body = json.loads(raw.decode("utf-8")) if raw else {}
        except Exception:
            body = {}

        # Legacy base64-wrapped payload support
        if "response" in body and isinstance(body["response"], str):
            try:
                decoded = base64.b64decode(body["response"]).decode("utf-8")
                body = json.loads(decoded)
            except Exception:
                pass

        event = body.get("event", "")
        payload = body.get("payload") or body.get("data") or {}
        merchant_order_id = payload.get("merchantOrderId") or payload.get("merchantTransactionId")
        state = (payload.get("state") or "").upper()

        if not merchant_order_id:
            return {"status": "ignored"}

        is_success = state == "COMPLETED" or "completed" in event.lower() or "success" in event.lower()
        is_failure = state in ("FAILED", "CANCELLED") or "failed" in event.lower() or "cancel" in event.lower()

        if is_success:
            await db.transactions.update_one(
                {"merchant_order_id": merchant_order_id},
                {"$set": {"payment_state": "COMPLETED", "updated_at": datetime.now(timezone.utc)}},
            )
            await db.appointments.update_one(
                {"transaction_id": merchant_order_id},
                {"$set": {"status": "confirmed", "payment_status": "paid"}},
            )
        elif is_failure:
            await db.transactions.update_one(
                {"merchant_order_id": merchant_order_id},
                {"$set": {"payment_state": "FAILED", "updated_at": datetime.now(timezone.utc)}},
            )
            await db.appointments.update_one(
                {"transaction_id": merchant_order_id},
                {"$set": {"status": "cancelled", "payment_status": "failed"}},
            )
        return {"status": "received"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Webhook error: {e}")
        return {"status": "error"}

File: backend-node/src/routes/payments.ts
/**
 * Payment routes: initiate, status, simulate, webhook (PhonePe v2).
 * Equivalent of Python's routes/payments.py
 */
import { Router, Request, Response } from 'express';
import { ObjectId } from 'mongodb';
import { v4 as uuidv4 } from 'uuid';
import axios from 'axios';
import { db } from '../core/database';
import { requireAuth } from '../core/middleware';
import { validate, PaymentInitiateSchema } from '../core/schemas';
import { logActivity } from '../services/activity';
import { handlePaymentSuccessNotification, handlePaymentFailureNotification } from '../services/email';
import { getPhonepeToken, verifyWebhookAuth, webhookAuthRequired } from '../services/phonepe';
import { FRONTEND_URL, PHONEPE_ENV, PHONEPE_PAY_URL, PHONEPE_STATUS_BASE, PHONEPE_MERCHANT_ID } from '../core/config';

const router = Router();

const completeAppointmentPayment = async (txnId: string, amountPaise: number): Promise<void> => {
  await db.appointments().updateOne(
    { transaction_id: txnId },
    { $set: { status: 'confirmed', payment_status: 'paid' } }
  );
  const apptForEmail = await db.appointments().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
  if (apptForEmail) {
    handlePaymentSuccessNotification({
      userEmail: (apptForEmail['patient_email'] as string) ?? '',
      userName: (apptForEmail['patient_name'] as string) ?? '',
      userPhone: (apptForEmail['patient_phone'] as string) ?? '',
      doctorName: (apptForEmail['doctor_name'] as string) ?? '',
      serviceName: (apptForEmail['service_name'] as string) ?? null,
      appointmentDate: (apptForEmail['appointment_date'] as string) ?? '',
      appointmentTime: (apptForEmail['appointment_time'] as string) ?? '',
      durationMinutes: (apptForEmail['duration_minutes'] as number) ?? 30,
      consultationFee: amountPaise / 100,
      transactionId: txnId,
    }).catch((err) => console.error('[Email] Notification failed:', err));
  }
};

const completeWorkshopPayment = async (txn: Record<string, unknown>): Promise<void> => {
  const registrationId = txn['workshop_registration_id'] as string | undefined;
  if (!registrationId) return;
  await db.workshop_registrations().updateOne(
    { _id: new ObjectId(registrationId) },
    {
      $set: {
        payment_status: 'paid',
        transaction_id: txn['transaction_id'],
        show_meet_link: true,
        updated_at: new Date(),
      },
    }
  );
};

// ─── POST /payments/initiate ──────────────────────────────────────────────────

router.post('/initiate', requireAuth, validate(PaymentInitiateSchema), async (req: Request, res: Response) => {
  const user = req.user!;
  const { appointment_id, workshop_registration_id } = req.body;

  let paymentSource: Record<string, unknown> | null = null;
  let entityType: 'appointment' | 'workshop_registration' = 'appointment';
  let amountRupees = 2000;
  let paymentLabel = '';
  let doctorName = '';
  let eventDate = '';
  let eventTime = '';

  if (appointment_id) {
    try {
      paymentSource = await db.appointments().findOne({ _id: new ObjectId(appointment_id) }) as Record<string, unknown> | null;
    } catch {
      res.status(404).json({ detail: 'Appointment not found' }); return;
    }
    if (!paymentSource) { res.status(404).json({ detail: 'Appointment not found' }); return; }
    if (paymentSource['user_id'] !== user['_id']) { res.status(403).json({ detail: 'Access denied' }); return; }
    amountRupees = (paymentSource['consultation_fee'] as number) ?? 2000;
    doctorName = (paymentSource['doctor_name'] as string) ?? '';
    eventDate = (paymentSource['appointment_date'] as string) ?? '';
    eventTime = (paymentSource['appointment_time'] as string) ?? '';
    paymentLabel = `Consultation fee - ${doctorName}`;
  } else {
    entityType = 'workshop_registration';
    try {
      paymentSource = await db.workshop_registrations().findOne({ _id: new ObjectId(workshop_registration_id) }) as Record<string, unknown> | null;
    } catch {
      res.status(404).json({ detail: 'Workshop registration not found' }); return;
    }
    if (!paymentSource) { res.status(404).json({ detail: 'Workshop registration not found' }); return; }
    if (paymentSource['user_id'] !== user['_id']) { res.status(403).json({ detail: 'Access denied' }); return; }
    amountRupees = (paymentSource['price'] as number) ?? 0;
    doctorName = (paymentSource['doctor_name'] as string) ?? '';
    eventDate = (paymentSource['workshop_date'] as string) ?? '';
    eventTime = (paymentSource['start_time'] as string) ?? '';
    paymentLabel = `Workshop fee - ${paymentSource['title'] ?? doctorName}`;
  }

  const amountPaise = Math.round(amountRupees * 100);
  const merchantOrderId = `MC${uuidv4().replace(/-/g, '').slice(0, 14).toUpperCase()}`;
  const txnId = merchantOrderId;
  const redirectUrl = `${FRONTEND_URL}/payment/status?txnId=${txnId}`;

  let checkoutUrl: string | null = null;
  let apiError: string | null = null;

  const accessToken = await getPhonepeToken();
  if (accessToken) {
    try {
      const payload = {
        merchantId: PHONEPE_MERCHANT_ID,
        merchantOrderId,
        amount: amountPaise,
        expireAfter: 1200,
        metaInfo: {
          udf1: user['_id'],
          udf2: doctorName,
          udf3: String(paymentSource['_id']),
        },
        paymentFlow: {
          type: 'PG_CHECKOUT',
          message: paymentLabel,
          merchantUrls: { redirectUrl },
        },
      };
      const resp = await axios.post(PHONEPE_PAY_URL, payload, {
        headers: { 'Content-Type': 'application/json', Authorization: `O-Bearer ${accessToken}` },
        timeout: 10000,
        validateStatus: () => true,
      });
      if (resp.status === 200 || resp.status === 201) {
        checkoutUrl = resp.data?.redirectUrl ?? resp.data?.data?.redirectUrl ?? null;
      } else {
        apiError = `${resp.status}: ${JSON.stringify(resp.data).slice(0, 200)}`;
        console.warn('[PhonePe] pay init failed:', apiError);
      }
    } catch (e: any) {
      apiError = String(e.message);
      console.warn('[PhonePe] API exception:', e);
    }
  }

  if (!checkoutUrl) {
    checkoutUrl = `${FRONTEND_URL}/payment/simulate/${txnId}`;
  }

  await db.transactions().insertOne({
    appointment_id: appointment_id ? String(paymentSource['_id']) : null,
    workshop_registration_id: workshop_registration_id ? String(paymentSource['_id']) : null,
    entity_type: entityType,
    user_id: user['_id'] as string,
    merchant_order_id: merchantOrderId,
    transaction_id: txnId,
    amount: amountPaise,
    payment_state: 'PENDING',
    checkout_url: checkoutUrl,
    doctor_name: doctorName,
    appointment_date: eventDate,
    appointment_time: eventTime,
    phonepe_env: PHONEPE_ENV,
    api_error: apiError,
    created_at: new Date(),
    updated_at: new Date(),
  });

  if (appointment_id) {
    await db.appointments().updateOne(
      { _id: new ObjectId(appointment_id) },
      { $set: { transaction_id: txnId } }
    );
  } else {
    await db.workshop_registrations().updateOne(
      { _id: new ObjectId(workshop_registration_id) },
      { $set: { transaction_id: txnId, updated_at: new Date() } }
    );
  }

  res.json({
    checkout_url: checkoutUrl,
    transaction_id: txnId,
    merchant_order_id: merchantOrderId,
    is_simulation: checkoutUrl.includes('/simulate/'),
  });
});

// ─── GET /payments/status/:txnId ─────────────────────────────────────────────

router.get('/status/:txnId', async (req: Request, res: Response) => {
  const { txnId } = req.params;
  const txn = await db.transactions().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
  if (!txn) { res.status(404).json({ detail: 'Transaction not found' }); return; }

  const safeT: Record<string, unknown> = { ...txn, _id: String(txn['_id']) };

  if (txn['payment_state'] === 'PENDING') {
    const accessToken = await getPhonepeToken();
    if (accessToken) {
      try {
        const merchantId = PHONEPE_MERCHANT_ID || txn['merchant_order_id']; 
        const url = `${PHONEPE_STATUS_BASE}/${txn['merchant_order_id']}/status?details=false&errorContext=true`;
        const resp = await axios.get(url, {
          headers: { 'Content-Type': 'application/json', Authorization: `O-Bearer ${accessToken}` },
          timeout: 10000,
          validateStatus: () => true,
        });
        if (resp.status === 200) {
          const newState: string = resp.data?.state ?? resp.data?.data?.state ?? '';
          if (['COMPLETED', 'FAILED'].includes(newState)) {
            await db.transactions().updateOne(
              { transaction_id: txnId },
              { $set: { payment_state: newState, updated_at: new Date() } }
            );
            safeT['payment_state'] = newState;
            if (newState === 'COMPLETED') {
              if (txn['entity_type'] === 'workshop_registration') {
                await completeWorkshopPayment({ ...txn, transaction_id: txnId });
              } else {
                await completeAppointmentPayment(txnId, (txn['amount'] as number) ?? 0);
              }
            } else {
              if (txn['entity_type'] === 'workshop_registration') {
                const registrationId = txn['workshop_registration_id'] as string | undefined;
                if (registrationId) {
                  await db.workshop_registrations().updateOne(
                    { _id: new ObjectId(registrationId) },
                    { $set: { payment_status: 'failed', show_meet_link: false, updated_at: new Date() } }
                  );
                }
              } else {
                await db.appointments().updateOne(
                  { transaction_id: txnId },
                  { $set: { status: 'cancelled', payment_status: 'failed' } }
                );
              }
              const failAppt = await db.appointments().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
              if (failAppt) {
                handlePaymentFailureNotification({
                  userEmail: (failAppt['patient_email'] as string) ?? '',
                  userName: (failAppt['patient_name'] as string) ?? '',
                  userPhone: (failAppt['patient_phone'] as string) ?? '',
                  doctorName: (failAppt['doctor_name'] as string) ?? '',
                  appointmentDate: (failAppt['appointment_date'] as string) ?? '',
                  appointmentTime: (failAppt['appointment_time'] as string) ?? '',
                  consultationFee: ((txn['amount'] as number) ?? 0) / 100,
                  transactionId: txnId,
                  paymentState: 'FAILED',
                }).catch((err) => console.error('[Email] Failure notification failed:', err));
              }
            }
          }
        }
      } catch (e) {
        console.warn('[PhonePe] status check failed:', e);
      }
    }
  }
  res.json(safeT);
});

// ─── POST /payments/simulate/:txnId/success ───────────────────────────────────

router.post('/simulate/:txnId/success', async (req: Request, res: Response) => {
  const { txnId } = req.params;
  const txn = await db.transactions().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
  if (!txn) { res.status(404).json({ detail: 'Transaction not found' }); return; }

  const provRef = `SIM${uuidv4().slice(0, 8).toUpperCase()}`;
  await db.transactions().updateOne(
    { transaction_id: txnId },
    {
      $set: {
        payment_state: 'COMPLETED',
        payment_mode: 'UPI_SIMULATE',
        provider_reference_id: provRef,
        updated_at: new Date(),
      },
    }
  );
  const appt = await db.appointments().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
  if (txn['entity_type'] === 'workshop_registration') {
    await completeWorkshopPayment({ ...txn, transaction_id: txnId });
    await logActivity(
      txn['user_id'] as string,
      'Patient',
      'WORKSHOP_PAYMENT_SUCCESS',
      `Workshop registration confirmed for ${txn['doctor_name'] ?? ''}`
    );
  } else if (appt) {
    await logActivity(
      txn['user_id'] as string,
      'Patient',
      'PAYMENT_SUCCESS',
      `Appointment confirmed for ${appt['doctor_name'] ?? ''}`
    );
    await completeAppointmentPayment(txnId, (txn['amount'] as number) ?? 0);
  }
  res.json({ message: 'Payment simulated successfully', transaction_id: txnId, provider_reference_id: provRef });
});

// ─── POST /payments/simulate/:txnId/failure ───────────────────────────────────

router.post('/simulate/:txnId/failure', async (req: Request, res: Response) => {
  const { txnId } = req.params;
  const txn = await db.transactions().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
  if (!txn) { res.status(404).json({ detail: 'Transaction not found' }); return; }

  await db.transactions().updateOne(
    { transaction_id: txnId },
    { $set: { payment_state: 'FAILED', updated_at: new Date() } }
  );
  if (txn['entity_type'] === 'workshop_registration') {
    const registrationId = txn['workshop_registration_id'] as string | undefined;
    if (registrationId) {
      await db.workshop_registrations().updateOne(
        { _id: new ObjectId(registrationId) },
        { $set: { payment_status: 'failed', show_meet_link: false, updated_at: new Date() } }
      );
    }
  } else {
    await db.appointments().updateOne(
      { transaction_id: txnId },
      { $set: { status: 'cancelled', payment_status: 'failed' } }
    );
  }
  const failAppt = await db.appointments().findOne({ transaction_id: txnId }) as Record<string, unknown> | null;
  if (failAppt) {
    handlePaymentFailureNotification({
      userEmail: (failAppt['patient_email'] as string) ?? '',
      userName: (failAppt['patient_name'] as string) ?? '',
      userPhone: (failAppt['patient_phone'] as string) ?? '',
      doctorName: (failAppt['doctor_name'] as string) ?? '',
      appointmentDate: (failAppt['appointment_date'] as string) ?? '',
      appointmentTime: (failAppt['appointment_time'] as string) ?? '',
      consultationFee: ((txn['amount'] as number) ?? 0) / 100,
      transactionId: txnId,
      paymentState: 'FAILED',
    }).catch((err) => console.error('[Email] Failure notification failed:', err));
  }
  res.json({ message: 'Payment failed', transaction_id: txnId });
});

// ─── POST /payments/webhook ───────────────────────────────────────────────────

router.post('/webhook', async (req: Request, res: Response) => {
  const authHeader = req.headers.authorization ?? '';
  if (webhookAuthRequired() && !verifyWebhookAuth(authHeader)) {
    console.warn('[PhonePe] Webhook rejected: invalid Authorization signature');
    res.status(401).json({ detail: 'Invalid webhook signature' });
    return;
  }
  if (!webhookAuthRequired()) {
    console.warn('[PhonePe] Webhook auth DISABLED (dev mode). Set PHONEPE_WEBHOOK_USERNAME/PASSWORD before production.');
  }

  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let body: Record<string, any> = req.body ?? {};

    // Legacy base64-wrapped payload support
    if (body['response'] && typeof body['response'] === 'string') {
      try {
        const decoded = Buffer.from(body['response'], 'base64').toString('utf-8');
        body = JSON.parse(decoded);
      } catch { /* ignore */ }
    }

    const event: string = body['event'] ?? '';
    const payload: Record<string, unknown> = body['payload'] ?? body['data'] ?? {};
    const merchantOrderId: string =
      (payload['merchantOrderId'] as string) ?? (payload['merchantTransactionId'] as string) ?? '';
    const state = String(payload['state'] ?? '').toUpperCase();

    if (!merchantOrderId) { res.json({ status: 'ignored' }); return; }

    const isSuccess =
      state === 'COMPLETED' ||
      event.toLowerCase().includes('completed') ||
      event.toLowerCase().includes('success');
    const isFailure =
      ['FAILED', 'CANCELLED'].includes(state) ||
      event.toLowerCase().includes('failed') ||
      event.toLowerCase().includes('cancel');

    if (isSuccess) {
      await db.transactions().updateOne(
        { merchant_order_id: merchantOrderId },
        { $set: { payment_state: 'COMPLETED', updated_at: new Date() } }
      );
      const whTxn = await db.transactions().findOne({ merchant_order_id: merchantOrderId }) as Record<string, unknown> | null;
      if (whTxn?.['entity_type'] === 'workshop_registration') {
        await completeWorkshopPayment({ ...whTxn, transaction_id: merchantOrderId });
      } else {
        await completeAppointmentPayment(merchantOrderId, (whTxn?.['amount'] as number) ?? 0);
      }
    } else if (isFailure) {
      await db.transactions().updateOne(
        { merchant_order_id: merchantOrderId },
        { $set: { payment_state: 'FAILED', updated_at: new Date() } }
      );
      const whFailTxn = await db.transactions().findOne({ merchant_order_id: merchantOrderId }) as Record<string, unknown> | null;
      if (whFailTxn?.['entity_type'] === 'workshop_registration') {
        const registrationId = whFailTxn['workshop_registration_id'] as string | undefined;
        if (registrationId) {
          await db.workshop_registrations().updateOne(
            { _id: new ObjectId(registrationId) },
            { $set: { payment_status: 'failed', show_meet_link: false, updated_at: new Date() } }
          );
        }
      } else {
        await db.appointments().updateOne(
          { transaction_id: merchantOrderId },
          { $set: { status: 'cancelled', payment_status: 'failed' } }
        );
      }
      const whFailAppt = await db.appointments().findOne({ transaction_id: merchantOrderId }) as Record<string, unknown> | null;
      if (whFailAppt) {
        handlePaymentFailureNotification({
          userEmail: (whFailAppt['patient_email'] as string) ?? '',
          userName: (whFailAppt['patient_name'] as string) ?? '',
          userPhone: (whFailAppt['patient_phone'] as string) ?? '',
          doctorName: (whFailAppt['doctor_name'] as string) ?? '',
          appointmentDate: (whFailAppt['appointment_date'] as string) ?? '',
          appointmentTime: (whFailAppt['appointment_time'] as string) ?? '',
          consultationFee: ((whFailTxn?.['amount'] as number) ?? 0) / 100,
          transactionId: merchantOrderId,
          paymentState: 'FAILED',
        }).catch((err) => console.error('[Email] Webhook failure notification failed:', err));
      }
    }
    res.json({ status: 'received' });
  } catch (e) {
    console.error('[PhonePe] Webhook error:', e);
    res.json({ status: 'error' });
  }
});

export default router;

